### PR TITLE
predix batch users changes

### PR DESCRIPTION
**Pull request by anny1099**

### CODE REVIEW

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java [hunk 1]
```diff
@@ -157,6 +157,15 @@ public ScimGroup createOrGet(ScimGroup group, String zoneId) {
         }
     }
 
+    @Override
+    public ScimGroup getOrCreate(ScimGroup group, String zoneId) {
```
Comment (irinaepshteyn) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:161]: So we have two implementations: createOrGet and getOrCreate
In the upstream PR https://github.com/cloudfoundry/uaa/pull/668, JdbcScimGroupMembershipManager was changed to use getOrCreate:
https://github.com/GESoftware-CF/uaa/blob/9ec558896229a19d47de455485ec905b1f7b3d8a/server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupMembershipManager.java#L110

Could you confirm which one should be used?

**Review by irinaepshteyn — CHANGES_REQUESTED** [review]
Also, missing changes to JdbcScimUserProvisioningTests from this commit: https://github.com/GESoftware-CF/uaa/commit/c1f56f48fa715a67a346a09425d837613e37f9cd

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java (hunk 1 above)
Comment (anny1099) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:161]: checked both upstream and pe_4.30.0 are using createOrGet in JdbcScimGroupMembershipManager.java. both getOrCreate and createOrGet has similar logic, the difference is that createOrGet throws exception of IncorrectResultSizeDataAccessException when there is error; vs. getOrCreate throws EmptyResultDataAccessException or ScimResourceAlreadyExistsException in the error cases.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserProvisioning.java [hunk 2]
```diff
@@ -25,6 +25,10 @@
 
     ScimUser createUser(ScimUser user, String password, String zoneId) throws InvalidPasswordException, InvalidScimResourceException;
 
+    default ScimUser createUser(ScimUser user, String password, boolean isBatchCall, String zoneId) throws InvalidPasswordException, InvalidScimResourceException {
```
[the remark this answers is not in the document]
Comment (anny1099) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserProvisioning.java]: done.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpoints.java [hunk 3]
```diff
@@ -225,15 +234,14 @@ public ScimUser createUser(@RequestBody ScimUser user, HttpServletRequest reques
             passwordValidator.validate(user.getPassword());
         }
 
-        ScimUser scimUser = scimUserProvisioning.createUser(user, user.getPassword(), identityZoneManager.getCurrentIdentityZoneId());
+        ScimUser scimUser = scimUserProvisioning.createUser(user, user.getPassword(), isBatchCall, identityZoneManager.getCurrentIdentityZoneId());
```
[the remark this answers is not in the document]
Comment (anny1099) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpoints.java]: done

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 4]
```diff
@@ -243,6 +267,13 @@ public ScimUser createUser(ScimUser user, final String password, String zoneId)
         return create(user, zoneId);
     }
 
+    @Override
+    public ScimUser createUser(ScimUser user, final String password, boolean isBatchCall, String zoneId) throws InvalidPasswordException,
```
[the remark this answers is not in the document]
Comment (anny1099) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: done

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpointDocs.java [hunk 5]
```diff
@@ -101,6 +102,7 @@
     private final String userLastLogonTimeDescription = "The unix epoch timestamp in milliseconds of when the user last authenticated. This field will be omitted from the response if the user has never authenticated.";
     private final String userPreviousLogonTimeDescription = "The unix epoch timestamp in milliseconds of 2nd to last successful user authentication. This field will only be included in the response once the user has authenticated two or more times.";
 
+    private final String scimWriteOrUaaAdminRequired = "Access token with `scim.write` or `uaa.admin` required";
```
[the remark this answers is not in the document]
Comment (anny1099) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpointDocs.java]: fixed.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ScimUserEndpointsIntegrationTests.java [hunk 6]
```diff
@@ -97,22 +101,43 @@ public void handleError(ClientHttpResponse response) {
...
 
-        return client.postForEntity(serverRunning.getUrl(userEndpoint), user, ScimUser.class);
+    private ResponseEntity<Object> createUserTx(String[] usernames, String[] firstNames, String[] lastNames, String[] emails) {
+        assertTrue("argument lengths not equal",
+                (usernames.length == firstNames.length) &&
+                (firstNames.length == lastNames.length) &&
+                (lastNames.length == emails.length));
+        ArrayList<ScimUser> users = new ArrayList<ScimUser>(usernames.length);
+        for(int i = 0; i < usernames.length; i++) {
+            ScimUser user = new ScimUser();
```
[the remark this answers is not in the document]
Comment (anny1099) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ScimUserEndpointsIntegrationTests.java]: done.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 7]
```diff
@@ -176,15 +177,31 @@ protected String getTableName() {
...
+        return super.query(filter, sortBy, ascending, zoneId);
+    }
+
     @Override
     public ScimUser create(final ScimUser user, String zoneId) {
+            return create(user, false, zoneId);
+    }
+
+    @Override
+    public ScimUser create(final ScimUser user, boolean isBatchCall, String zoneId) {
```
[the remark this answers is not in the document]
Comment (anny1099) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: done.

**Comment by anny1099** [comment]
> Also, missing changes to JdbcScimUserProvisioningTests from this commit: [c1f56f4](https://github.com/GESoftware-CF/uaa/commit/c1f56f48fa715a67a346a09425d837613e37f9cd)

fixed.

**Review by irinaepshteyn — APPROVED** [review]
@anny1099, done reviewing.  Approved

**Comment by irinaepshteyn** [comment]
@anny1099, before merging, could you please check why ScimUserEndpointsIntegrationTests. createUserTxInvalid integration test is failing in Jenkins

**Comment by anny1099** [comment]
build passed.https://i.ci.build.ge.com/bjmjek1k/ci/job/UAA/job/Build%20n%20Test/job/port_users_changes/10/

**Review by irinaepshteyn — APPROVED** [review]

**Pull request merged**